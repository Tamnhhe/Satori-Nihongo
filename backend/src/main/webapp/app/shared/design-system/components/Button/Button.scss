@import '../../tokens/variables';

.satori-button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-family: $font-family-sans;
  font-weight: $font-weight-medium;
  border-radius: $radius-md;
  border: 1px solid transparent;
  cursor: pointer;
  transition: all $transition-fast;
  text-decoration: none;
  white-space: nowrap;
  user-select: none;
  position: relative;
  overflow: hidden;

  &:focus {
    outline: none;
    box-shadow: 0 0 0 3px rgba($primary-500, 0.1);
  }

  &:active {
    transform: translateY(1px);
  }

  // Size variants
  &--sm {
    height: $button-height-sm;
    padding: 0 $spacing-3;
    font-size: $text-sm;
    line-height: $leading-tight;

    &.satori-button--icon-only {
      width: $button-height-sm;
      padding: 0;
    }
  }

  &--md {
    height: $button-height-md;
    padding: 0 $spacing-4;
    font-size: $text-base;
    line-height: $leading-normal;

    &.satori-button--icon-only {
      width: $button-height-md;
      padding: 0;
    }
  }

  &--lg {
    height: $button-height-lg;
    padding: 0 $spacing-6;
    font-size: $text-lg;
    line-height: $leading-normal;

    &.satori-button--icon-only {
      width: $button-height-lg;
      padding: 0;
    }
  }

  // Variant styles
  &--primary {
    background-color: $primary-600;
    color: white;
    border-color: $primary-600;

    &:hover:not(:disabled) {
      background-color: $primary-700;
      border-color: $primary-700;
    }

    &:active:not(:disabled) {
      background-color: $primary-800;
      border-color: $primary-800;
    }
  }

  &--secondary {
    background-color: $secondary-600;
    color: white;
    border-color: $secondary-600;

    &:hover:not(:disabled) {
      background-color: $secondary-700;
      border-color: $secondary-700;
    }

    &:active:not(:disabled) {
      background-color: $secondary-800;
      border-color: $secondary-800;
    }
  }

  &--outline {
    background-color: transparent;
    color: $gray-700;
    border-color: $gray-300;

    &:hover:not(:disabled) {
      background-color: $gray-50;
      border-color: $gray-400;
      color: $gray-900;
    }

    &:active:not(:disabled) {
      background-color: $gray-100;
    }
  }

  &--ghost {
    background-color: transparent;
    color: $gray-600;
    border-color: transparent;

    &:hover:not(:disabled) {
      background-color: $gray-100;
      color: $gray-900;
    }

    &:active:not(:disabled) {
      background-color: $gray-200;
    }
  }

  &--danger {
    background-color: $error-600;
    color: white;
    border-color: $error-600;

    &:hover:not(:disabled) {
      background-color: $error-700;
      border-color: $error-700;
    }

    &:active:not(:disabled) {
      background-color: $error-800;
      border-color: $error-800;
    }
  }

  &--success {
    background-color: $success-600;
    color: white;
    border-color: $success-600;

    &:hover:not(:disabled) {
      background-color: $success-700;
      border-color: $success-700;
    }

    &:active:not(:disabled) {
      background-color: $success-800;
      border-color: $success-800;
    }
  }

  &--warning {
    background-color: $warning-600;
    color: white;
    border-color: $warning-600;

    &:hover:not(:disabled) {
      background-color: $warning-700;
      border-color: $warning-700;
    }

    &:active:not(:disabled) {
      background-color: $warning-800;
      border-color: $warning-800;
    }
  }

  // States
  &--disabled,
  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
    pointer-events: none;
  }

  &--loading {
    cursor: wait;
    pointer-events: none;
  }

  &--full-width {
    width: 100%;
  }

  // Content and icons
  &__content {
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &__icon {
    flex-shrink: 0;

    &--left {
      margin-right: $spacing-2;
    }

    &--right {
      margin-left: $spacing-2;
    }

    &--spinning {
      animation: spin 1s linear infinite;
    }
  }

  // Icon only buttons
  &--icon-only {
    .satori-button__icon {
      margin: 0;
    }
  }
}

// Button Group
.satori-button-group {
  display: flex;

  &--horizontal {
    flex-direction: row;

    &.satori-button-group--spacing-none {
      .satori-button:not(:first-child) {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        margin-left: -1px;
      }

      .satori-button:not(:last-child) {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }
    }

    &.satori-button-group--spacing-sm {
      gap: $spacing-2;
    }

    &.satori-button-group--spacing-md {
      gap: $spacing-3;
    }

    &.satori-button-group--spacing-lg {
      gap: $spacing-4;
    }
  }

  &--vertical {
    flex-direction: column;

    &.satori-button-group--spacing-none {
      .satori-button:not(:first-child) {
        border-top-left-radius: 0;
        border-top-right-radius: 0;
        margin-top: -1px;
      }

      .satori-button:not(:last-child) {
        border-bottom-left-radius: 0;
        border-bottom-right-radius: 0;
      }
    }

    &.satori-button-group--spacing-sm {
      gap: $spacing-2;
    }

    &.satori-button-group--spacing-md {
      gap: $spacing-3;
    }

    &.satori-button-group--spacing-lg {
      gap: $spacing-4;
    }
  }
}

// Animations
@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
