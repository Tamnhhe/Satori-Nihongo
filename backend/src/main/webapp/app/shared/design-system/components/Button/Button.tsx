import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { useKeyboardNavigation, useAriaAttributes } from '../../../hooks/useAccessibility';
import './Button.scss';

export interface ButtonProps {
  children: React.ReactNode;
  variant?: 'primary' | 'secondary' | 'outline' | 'ghost' | 'danger' | 'success' | 'warning';
  size?: 'sm' | 'md' | 'lg';
  icon?: IconDefinition;
  iconPosition?: 'left' | 'right';
  loading?: boolean;
  disabled?: boolean;
  fullWidth?: boolean;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  onKeyDown?: (event: React.KeyboardEvent<HTMLButtonElement>) => void;
  type?: 'button' | 'submit' | 'reset';
  className?: string;
  'aria-label'?: string;
  'aria-describedby'?: string;
  'aria-expanded'?: boolean;
  'aria-pressed'?: boolean;
  'aria-current'?: boolean | 'page' | 'step' | 'location' | 'date' | 'time';
  id?: string;
  role?: string;
  tabIndex?: number;
  autoFocus?: boolean;
}

export const Button: React.FC<ButtonProps> = ({
  children,
  variant = 'primary',
  size = 'md',
  icon,
  iconPosition = 'left',
  loading = false,
  disabled = false,
  fullWidth = false,
  onClick,
  onKeyDown,
  type = 'button',
  className = '',
  'aria-label': ariaLabel,
  'aria-describedby': ariaDescribedBy,
  'aria-expanded': ariaExpanded,
  'aria-pressed': ariaPressed,
  'aria-current': ariaCurrent,
  id,
  role,
  tabIndex,
  autoFocus = false,
}) => {
  const buttonClasses = [
    'satori-button',
    `satori-button--${variant}`,
    `satori-button--${size}`,
    fullWidth && 'satori-button--full-width',
    loading && 'satori-button--loading',
    disabled && 'satori-button--disabled',
    className,
  ]
    .filter(Boolean)
    .join(' ');

  const iconElement = icon && (
    <FontAwesomeIcon
      icon={loading ? faSpinner : icon}
      className={`satori-button__icon satori-button__icon--${iconPosition} ${loading ? 'satori-button__icon--spinning' : ''}`}
      aria-hidden="true"
    />
  );

  // Enhanced keyboard navigation
  const { handleKeyDown: handleKeyboardNav } = useKeyboardNavigation(() => onClick && onClick({} as React.MouseEvent<HTMLButtonElement>));

  const handleKeyDownEvent = (event: React.KeyboardEvent<HTMLButtonElement>) => {
    if (onKeyDown) {
      onKeyDown(event);
    }
    if (!event.defaultPrevented) {
      handleKeyboardNav(event.nativeEvent);
    }
  };

  // ARIA attributes
  const ariaAttributes = useAriaAttributes(
    role,
    ariaLabel || (typeof children === 'string' ? children : undefined),
    undefined,
    ariaDescribedBy,
    ariaExpanded,
    undefined,
    disabled || loading,
    undefined,
    undefined,
    ariaCurrent,
  );

  // Additional ARIA attributes
  if (ariaPressed !== undefined) {
    ariaAttributes['aria-pressed'] = ariaPressed;
  }

  return (
    <button
      id={id}
      type={type}
      className={buttonClasses}
      onClick={onClick}
      onKeyDown={handleKeyDownEvent}
      disabled={disabled || loading}
      tabIndex={tabIndex}
      autoFocus={autoFocus}
      {...ariaAttributes}
    >
      {loading && <span className="sr-only">Loading...</span>}
      {iconPosition === 'left' && iconElement}
      <span className="satori-button__content">{children}</span>
      {iconPosition === 'right' && iconElement}
    </button>
  );
};

export interface IconButtonProps {
  icon: IconDefinition;
  variant?: 'primary' | 'secondary' | 'outline' | 'ghost' | 'danger';
  size?: 'sm' | 'md' | 'lg';
  loading?: boolean;
  disabled?: boolean;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  onKeyDown?: (event: React.KeyboardEvent<HTMLButtonElement>) => void;
  type?: 'button' | 'submit' | 'reset';
  className?: string;
  'aria-label': string;
  'aria-describedby'?: string;
  'aria-expanded'?: boolean;
  'aria-pressed'?: boolean;
  'aria-current'?: boolean | 'page' | 'step' | 'location' | 'date' | 'time';
  id?: string;
  role?: string;
  tabIndex?: number;
  autoFocus?: boolean;
}

export const IconButton: React.FC<IconButtonProps> = ({
  icon,
  variant = 'ghost',
  size = 'md',
  loading = false,
  disabled = false,
  onClick,
  onKeyDown,
  type = 'button',
  className = '',
  'aria-label': ariaLabel,
  'aria-describedby': ariaDescribedBy,
  'aria-expanded': ariaExpanded,
  'aria-pressed': ariaPressed,
  'aria-current': ariaCurrent,
  id,
  role,
  tabIndex,
  autoFocus = false,
}) => {
  const buttonClasses = [
    'satori-button',
    'satori-button--icon-only',
    `satori-button--${variant}`,
    `satori-button--${size}`,
    loading && 'satori-button--loading',
    disabled && 'satori-button--disabled',
    'touch-target', // Ensure minimum touch target size
    className,
  ]
    .filter(Boolean)
    .join(' ');

  // Enhanced keyboard navigation
  const { handleKeyDown: handleKeyboardNav } = useKeyboardNavigation(() => onClick && onClick({} as React.MouseEvent<HTMLButtonElement>));

  const handleKeyDownEvent = (event: React.KeyboardEvent<HTMLButtonElement>) => {
    if (onKeyDown) {
      onKeyDown(event);
    }
    if (!event.defaultPrevented) {
      handleKeyboardNav(event.nativeEvent);
    }
  };

  // ARIA attributes
  const ariaAttributes = useAriaAttributes(
    role,
    ariaLabel,
    undefined,
    ariaDescribedBy,
    ariaExpanded,
    undefined,
    disabled || loading,
    undefined,
    undefined,
    ariaCurrent,
  );

  // Additional ARIA attributes
  if (ariaPressed !== undefined) {
    ariaAttributes['aria-pressed'] = ariaPressed;
  }

  return (
    <button
      id={id}
      type={type}
      className={buttonClasses}
      onClick={onClick}
      onKeyDown={handleKeyDownEvent}
      disabled={disabled || loading}
      tabIndex={tabIndex}
      autoFocus={autoFocus}
      {...ariaAttributes}
    >
      {loading && <span className="sr-only">Loading...</span>}
      <FontAwesomeIcon
        icon={loading ? faSpinner : icon}
        className={`satori-button__icon ${loading ? 'satori-button__icon--spinning' : ''}`}
        aria-hidden="true"
      />
    </button>
  );
};

export interface ButtonGroupProps {
  children: React.ReactNode;
  className?: string;
  orientation?: 'horizontal' | 'vertical';
  spacing?: 'none' | 'sm' | 'md' | 'lg';
}

export const ButtonGroup: React.FC<ButtonGroupProps> = ({ children, className = '', orientation = 'horizontal', spacing = 'sm' }) => {
  const groupClasses = ['satori-button-group', `satori-button-group--${orientation}`, `satori-button-group--spacing-${spacing}`, className]
    .filter(Boolean)
    .join(' ');

  return <div className={groupClasses}>{children}</div>;
};
