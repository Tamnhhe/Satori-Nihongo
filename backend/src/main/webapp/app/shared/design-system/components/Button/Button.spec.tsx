import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Button } from './Button';

expect.extend(toHaveNoViolations);

describe('Button Component', () => {
  it('should render with default props', () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole('button', { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('btn', 'btn-primary');
  });

  it('should render different variants correctly', () => {
    const { rerender } = render(<Button variant="secondary">Secondary</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-secondary');

    rerender(<Button variant="outline">Outline</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-outline');

    rerender(<Button variant="ghost">Ghost</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-ghost');
  });

  it('should render different sizes correctly', () => {
    const { rerender } = render(<Button size="small">Small</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-sm');

    rerender(<Button size="large">Large</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-lg');
  });

  it('should handle disabled state', () => {
    render(<Button disabled>Disabled</Button>);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    expect(button).toHaveClass('btn-disabled');
  });

  it('should handle loading state', () => {
    render(<Button loading>Loading</Button>);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    expect(button).toHaveClass('btn-loading');
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
  });

  it('should handle click events', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(<Button onClick={handleClick}>Click me</Button>);
    const button = screen.getByRole('button');

    await user.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('should not trigger click when disabled', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(
      <Button onClick={handleClick} disabled>
        Disabled
      </Button>,
    );
    const button = screen.getByRole('button');

    await user.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('should render with icon', () => {
    render(<Button icon="plus">Add Item</Button>);
    expect(screen.getByTestId('button-icon')).toBeInTheDocument();
  });

  it('should support keyboard navigation', async () => {
    const handleClick = jest.fn();
    const user = userEvent.setup();

    render(<Button onClick={handleClick}>Press Enter</Button>);
    const button = screen.getByRole('button');

    button.focus();
    await user.keyboard('{Enter}');
    expect(handleClick).toHaveBeenCalledTimes(1);

    await user.keyboard(' ');
    expect(handleClick).toHaveBeenCalledTimes(2);
  });

  it('should be accessible', async () => {
    const { container } = render(
      <Button aria-label="Accessible button">
        <span aria-hidden="true">ðŸ”¥</span>
        Action
      </Button>,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should have proper ARIA attributes', () => {
    render(
      <Button aria-describedby="help-text" aria-pressed={true} role="switch">
        Toggle
      </Button>,
    );

    const button = screen.getByRole('switch');
    expect(button).toHaveAttribute('aria-describedby', 'help-text');
    expect(button).toHaveAttribute('aria-pressed', 'true');
  });

  it('should match snapshot', () => {
    const { container } = render(
      <div>
        <Button variant="primary">Primary</Button>
        <Button variant="secondary" size="small">
          Secondary Small
        </Button>
        <Button variant="outline" disabled>
          Outline Disabled
        </Button>
        <Button loading>Loading</Button>
      </div>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
