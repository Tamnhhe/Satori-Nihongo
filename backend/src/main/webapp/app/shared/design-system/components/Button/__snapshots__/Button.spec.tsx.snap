// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should match snapshot 1`] = `
<div>
  <button
    aria-disabled="false"
    aria-label="Primary"
    class="satori-button satori-button--primary satori-button--md"
    type="button"
  >
    <span
      class="satori-button__content"
    >
      Primary
    </span>
  </button>
  <button
    aria-disabled="false"
    aria-label="Secondary Small"
    class="satori-button satori-button--secondary satori-button--small"
    type="button"
  >
    <span
      class="satori-button__content"
    >
      Secondary Small
    </span>
  </button>
  <button
    aria-disabled="true"
    aria-label="Outline Disabled"
    class="satori-button satori-button--outline satori-button--md satori-button--disabled"
    disabled=""
    type="button"
  >
    <span
      class="satori-button__content"
    >
      Outline Disabled
    </span>
  </button>
  <button
    aria-disabled="true"
    aria-label="Loading"
    class="satori-button satori-button--primary satori-button--md satori-button--loading"
    disabled=""
    type="button"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
    <span
      class="satori-button__content"
    >
      Loading
    </span>
  </button>
</div>
`;
